packages/@lwc/module-resolver/src/index.ts

```diff
@@ -4,117 +4,147 @@
+
+    if (modules) {
+        validateNpmConfig(lwcConfig);
+        const exposedModules = lwcConfig.expose || [];
+        if (exposedModules.includes(specifier)) {
+            for (const moduleRecord of modules) {
+                if (!isNpmModuleRecord(moduleRecord)) {
+                    const registryEntry = resolveModuleRecordType(specifier, moduleRecord, {
+                        rootDir: packageDir,
+                    });
+                    if (registryEntry) {
+                        if (aliasMapping && aliasMapping[specifier]) {
+                            registryEntry.specifier = aliasMapping[specifier];
+                        }
+                        return registryEntry;
+                    }
                 }
-            });
+            }
+
+            throw new Error(`Unable to find ${specifier} under package ${npmModuleRecord.npm}`);
         }
-    });
+    }
+}
 
-    return resolvedModules;
+function resolveModuleRecordType(
+    specifier: string,
+    moduleRecord: ModuleRecord,
+    opts: InnerResolverOptions
+): RegistryEntry | undefined {
+    const { rootDir } = opts;
+    validateModuleRecord(moduleRecord);
+
+    if (isAliasModuleRecord(moduleRecord) && moduleRecord.name === specifier) {
+        return resolveModuleFromAlias(moduleRecord, { rootDir });
+    } else if (isDirModuleRecord(moduleRecord)) {
+        return resolveModuleFromDir(specifier, moduleRecord, { rootDir });
+    } else if (isNpmModuleRecord(moduleRecord)) {
+        return resolveModuleFromNpm(specifier, moduleRecord, opts);
+    } else {
+        throw new Error(`Invalid moduleRecord type ${moduleRecord}`);
+    }
 }
 
-function resolveModulesFromNpm(packageName: string): RegistryEntry[] {
-    let resolvedModules: RegistryEntry[] = [];
-    try {
-        const pkgJsonPath = require.resolve(`${packageName}/package.json`);
-        const packageDir = path.dirname(pkgJsonPath);
-        const lwcConfigFile = path.join(packageDir, LWC_CONFIG_FILE);
-
-        if (fs.existsSync(lwcConfigFile)) {
-            resolvedModules = resolveModules({ rootDir: lwcConfigFile });
-        } else {
-            const pkgJson = JSON.parse(fs.readFileSync(pkgJsonPath, 'utf8'));
-            if (pkgJson.lwc) {
-                resolvedModules = resolveModulesFromList(pkgJson.lwc.modules, { root: packageDir });
-            }
-        }
-    } catch (e) {
-        /*noop*/
+export function resolveModule(
+    importee: string,
+    importer: string,
+    config?: Partial<ModuleResolverConfig>
+): RegistryEntry {
+    validateImportee(importee);
+
+    const rootDir = findFirstUpwardConfigPath(path.resolve(importer));
+    const lwcConfig = getLwcConfig(rootDir);
+    let modules: ModuleRecord[] = lwcConfig.modules || [];
+
+    if (config) {
+        const userConfig = normalizeConfig(config);
+        modules = mergeModules(userConfig.modules, modules);
     }
 
-    return resolvedModules;
-}
+    if (modules.length === 0) {
```

Comment:
No need for this check, you can just rely on the fact that it will throw right after.